File: apprefurbished/src/main/java/com/androidcollider/easyfin/transactions/add_edit/btw_accounts/AddTransactionBetweenAccountsFragment.java
package com.androidcollider.easyfin.transactions.add_edit.btw_accounts;

import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.RelativeLayout;
import android.widget.ScrollView;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.Nullable;

import com.androidcollider.easyfin.R;
import com.androidcollider.easyfin.common.app.App;
import com.androidcollider.easyfin.common.events.UpdateFrgAccounts;
import com.androidcollider.easyfin.common.events.UpdateFrgHomeBalance;
import com.androidcollider.easyfin.common.managers.resources.ResourcesManager;
import com.androidcollider.easyfin.common.managers.ui.hide_touch_outside.HideTouchOutsideManager;
import com.androidcollider.easyfin.common.managers.ui.shake_edit_text.ShakeEditTextManager;
import com.androidcollider.easyfin.common.managers.ui.toast.ToastManager;
import com.androidcollider.easyfin.common.ui.adapters.SpinAccountForTransAdapter;
import com.androidcollider.easyfin.common.ui.fragments.NumericDialogFragment;
import com.androidcollider.easyfin.common.ui.fragments.common.CommonFragmentAddEdit;
import com.androidcollider.easyfin.common.utils.EditTextAmountWatcher;
import com.androidcollider.easyfin.common.view_models.SpinAccountViewModel;

import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

/**
 * @author Ihor Bilous
 */

public class AddTransactionBetweenAccountsFragment extends CommonFragmentAddEdit
        implements NumericDialogFragment.OnCommitAmountListener, AddTransactionBetweenAccountsMVP.View {

    Spinner spinAccountFrom;
    Spinner spinAccountTo;
    EditText etExchange;
    TextView tvAmount;
    RelativeLayout layoutExchange;
    ScrollView scrollView;

    private SpinAccountForTransAdapter adapterAccountTo;
    private List<SpinAccountViewModel> accountListFrom, accountListTo;

    @Inject
    ShakeEditTextManager shakeEditTextManager;

    @Inject
    ToastManager toastManager;

    @Inject
    HideTouchOutsideManager hideTouchOutsideManager;

    @Inject
    ResourcesManager resourcesManager;

    @Inject
    AddTransactionBetweenAccountsMVP.Presenter presenter;


    @Override
    public int getContentView() {
        return R.layout.frg_add_trans_btw;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((App) getActivity().getApplication()).getComponent().inject(this);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupUI(view);
        setToolbar();
        accountListFrom = new ArrayList<>();

        presenter.setView(this);
        presenter.loadAccounts();
    }

    private void setupUI(View view) {
        spinAccountFrom = view.findViewById(R.id.spinAddTransBTWAccountFrom);
        spinAccountTo = view.findViewById(R.id.spinAddTransBTWAccountTo);
        etExchange = view.findViewById(R.id.editTextTransBTWExchange);
        tvAmount = view.findViewById(R.id.tvAddTransBTWAmount);
        layoutExchange = view.findViewById(R.id.layoutAddTransBTWExchange);
        scrollView = view.findViewById(R.id.scrollAddTransBTW);

        tvAmount.setOnClickListener(v -> openNumericDialog());
    }

    private void setSpinners() {
        accountListTo = new ArrayList<>();

        spinAccountFrom.setAdapter(new SpinAccountForTransAdapter(
                getActivity(),
                R.layout.spin_head_text,
                accountListFrom,
                resourcesManager
        ));

        accountListTo.addAll(accountListFrom);
        accountListTo.remove(spinAccountFrom.getSelectedItemPosition());

        adapterAccountTo = new SpinAccountForTransAdapter(
                getActivity(),
                R.layout.spin_head_text,
                accountListTo,
                resourcesManager
        );

        spinAccountTo.setAdapter(adapterAccountTo);

        spinAccountFrom.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                updateSpinnerTo();
                presenter.setCurrencyMode();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });

        spinAccountTo.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                presenter.setCurrencyMode();
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {
            }
        });
    }

    private void updateSpinnerTo() {
        accountListTo.clear();
        accountListTo.addAll(accountListFrom);
        accountListTo.remove(spinAccountFrom.getSelectedItemPosition());
        adapterAccountTo.notifyDataSetChanged();
    }

    private void pushBroadcast() {
        EventBus.getDefault().post(new UpdateFrgHomeBalance());
        EventBus.getDefault().post(new UpdateFrgAccounts());
    }

    @Override
    public void onCommitAmountSubmit(String amount) {
        showAmount(amount);
    }


    @Override
    public void handleSaveAction() {
        presenter.save();
    }

    @Override
    public void showAmount(String amount) {
        setTVTextSize(tvAmount, amount, 10, 15);
        tvAmount.setText(amount);
    }

    @Override
    public void showExchangeRate(String rate) {
        layoutExchange.setVisibility(View.VISIBLE);
        etExchange.setText(rate);
        etExchange.setSelection(etExchange.getText().length());
    }

    @Override
    public void hideExchangeRate() {
        layoutExchange.setVisibility(View.GONE);
    }

    @Override
    public void highlightExchangeRateField() {
        shakeEditTextManager.highlightEditText(etExchange);
    }

    @Override
    public void showMessage(String message) {
        toastManager.showClosableToast(getActivity(), message, ToastManager.SHORT);
    }

    @Override
    public void openNumericDialog() {
        openNumericDialog(tvAmount.getText().toString());
    }

    @Override
    public void performLastActionsAfterSaveAndClose() {
        pushBroadcast();
        finish();
    }

    @Override
    public String getAmount() {
        return tvAmount.getText().toString();
    }

    @Override
    public String getExchangeRate() {
        return etExchange.getText().toString();
    }

    @Override
    public SpinAccountViewModel getAccountFrom() {
        return (SpinAccountViewModel) spinAccountFrom.getSelectedItem();
    }

    @Override
    public SpinAccountViewModel getAccountTo() {
        return (SpinAccountViewModel) spinAccountTo.getSelectedItem();
    }

    @Override
    public boolean isMultiCurrencyTransaction() {
        return layoutExchange.getVisibility() == View.VISIBLE;
    }

    @Override
    public void notifyNotEnoughAccounts() {
        scrollView.setVisibility(View.GONE);
        showDialogNoAccount(getString(R.string.dialog_text_transfer_no_accounts), false);
    }

    @Override
    public void setAccounts(List<SpinAccountViewModel> accountList) {
        accountListFrom.clear();
        accountListFrom.addAll(accountList);

        scrollView.setVisibility(View.VISIBLE);
        showAmount("0,00");
        openNumericDialog();
        etExchange.addTextChangedListener(new EditTextAmountWatcher(etExchange));
        setSpinners();
        hideTouchOutsideManager.hideKeyboardByTouchOutsideEditText(scrollView, getActivity());
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/transactions/add_edit/btw_accounts/AddTransactionBetweenAccountsFragment.kt
package com.androidcollider.easyfin.transactions.add_edit.btw_accounts

import android.os.Bundle
import android.view.View
import android.widget.*
import com.androidcollider.easyfin.R
import com.androidcollider.easyfin.common.app.App
import com.androidcollider.easyfin.common.events.UpdateFrgAccounts
import com.androidcollider.easyfin.common.events.UpdateFrgHomeBalance
import com.androidcollider.easyfin.common.managers.resources.ResourcesManager
import com.androidcollider.easyfin.common.managers.ui.hide_touch_outside.HideTouchOutsideManager
import com.androidcollider.easyfin.common.managers.ui.shake_edit_text.ShakeEditTextManager
import com.androidcollider.easyfin.common.managers.ui.toast.ToastManager
import com.androidcollider.easyfin.common.ui.adapters.SpinAccountForTransAdapter
import com.androidcollider.easyfin.common.ui.fragments.NumericDialogFragment.OnCommitAmountListener
import com.androidcollider.easyfin.common.ui.fragments.common.CommonFragmentAddEdit
import com.androidcollider.easyfin.common.utils.EditTextAmountWatcher
import com.androidcollider.easyfin.common.view_models.SpinAccountViewModel
import org.greenrobot.eventbus.EventBus
import javax.inject.Inject

/**
 * @author Ihor Bilous
 */
class AddTransactionBetweenAccountsFragment : CommonFragmentAddEdit(), OnCommitAmountListener,
    AddTransactionBetweenAccountsMVP.View {
    private lateinit var spinAccountFrom: Spinner
    private lateinit var spinAccountTo: Spinner
    private lateinit var etExchange: EditText
    private lateinit var tvAmount: TextView
    private lateinit var layoutExchange: RelativeLayout
    private lateinit var scrollView: ScrollView
    private lateinit var adapterAccountTo: SpinAccountForTransAdapter
    private lateinit var accountListFrom: MutableList<SpinAccountViewModel>
    private lateinit var accountListTo: MutableList<SpinAccountViewModel>

    @Inject
    lateinit var shakeEditTextManager: ShakeEditTextManager

    @Inject
    lateinit var toastManager: ToastManager

    @Inject
    lateinit var hideTouchOutsideManager: HideTouchOutsideManager

    @Inject
    lateinit var resourcesManager: ResourcesManager

    @Inject
    lateinit var presenter: AddTransactionBetweenAccountsMVP.Presenter

    override val contentView: Int
        get() = R.layout.frg_add_trans_btw

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity?.application as App).component?.inject(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupUI(view)
        setToolbar()
        accountListFrom = ArrayList()
        presenter.setView(this)
        presenter.loadAccounts()
    }

    private fun setupUI(view: View) {
        spinAccountFrom = view.findViewById(R.id.spinAddTransBTWAccountFrom)
        spinAccountTo = view.findViewById(R.id.spinAddTransBTWAccountTo)
        etExchange = view.findViewById(R.id.editTextTransBTWExchange)
        tvAmount = view.findViewById(R.id.tvAddTransBTWAmount)
        layoutExchange = view.findViewById(R.id.layoutAddTransBTWExchange)
        scrollView = view.findViewById(R.id.scrollAddTransBTW)
        tvAmount.setOnClickListener { openNumericDialog() }
    }

    private fun setSpinners() {
        accountListTo = ArrayList()
        activity?.let {
            spinAccountFrom.adapter = SpinAccountForTransAdapter(
                it,
                R.layout.spin_head_text,
                accountListFrom,
                resourcesManager
            )
        }

        accountListTo.addAll(accountListFrom)
        accountListTo.removeAt(spinAccountFrom.selectedItemPosition)

        activity?.let {
            adapterAccountTo = SpinAccountForTransAdapter(
                it,
                R.layout.spin_head_text,
                accountListTo,
                resourcesManager
            )
        }

        spinAccountTo.adapter = adapterAccountTo
        spinAccountFrom.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View,
                position: Int,
                id: Long
            ) {
                updateSpinnerTo()
                presenter.setCurrencyMode()
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
        spinAccountTo.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>?, view: View, i: Int, l: Long) {
                presenter.setCurrencyMode()
            }

            override fun onNothingSelected(adapterView: AdapterView<*>?) {}
        }
    }

    private fun updateSpinnerTo() {
        accountListTo.clear()
        accountListTo.addAll(accountListFrom)
        accountListTo.removeAt(spinAccountFrom.selectedItemPosition)
        adapterAccountTo.notifyDataSetChanged()
    }

    private fun pushBroadcast() {
        EventBus.getDefault().post(UpdateFrgHomeBalance())
        EventBus.getDefault().post(UpdateFrgAccounts())
    }

    override fun onCommitAmountSubmit(amount: String) {
        showAmount(amount)
    }

    public override fun handleSaveAction() {
        presenter.save()
    }

    override fun showAmount(amount: String?) {
        setTVTextSize(tvAmount, amount!!, 10, 15)
        tvAmount.text = amount
    }

    override fun showExchangeRate(rate: String?) {
        layoutExchange.visibility = View.VISIBLE
        etExchange.setText(rate)
        etExchange.setSelection(etExchange.text.length)
    }

    override fun hideExchangeRate() {
        layoutExchange.visibility = View.GONE
    }

    override fun highlightExchangeRateField() {
        shakeEditTextManager.highlightEditText(etExchange)
    }

    override fun showMessage(message: String) {
        activity?.let {
            toastManager.showClosableToast(it, message, ToastManager.SHORT)
        }
    }

    override fun openNumericDialog() {
        openNumericDialog(tvAmount.text.toString())
    }

    override fun performLastActionsAfterSaveAndClose() {
        pushBroadcast()
        finish()
    }

    override val amount: String
        get() = tvAmount.text.toString()
    override val exchangeRate: String
        get() = etExchange.text.toString()
    override val accountFrom: SpinAccountViewModel
        get() = spinAccountFrom.selectedItem as SpinAccountViewModel
    override val accountTo: SpinAccountViewModel
        get() = spinAccountTo.selectedItem as SpinAccountViewModel
    override val isMultiCurrencyTransaction: Boolean
        get() = layoutExchange.visibility == View.VISIBLE

    override fun notifyNotEnoughAccounts() {
        scrollView.visibility = View.GONE
        showDialogNoAccount(getString(R.string.dialog_text_transfer_no_accounts), false)
    }

    override fun setAccounts(accountList: List<SpinAccountViewModel>?) {
        accountListFrom.clear()
        accountListFrom.addAll(accountList!!)
        scrollView.visibility = View.VISIBLE
        showAmount("0,00")
        openNumericDialog()
        etExchange.addTextChangedListener(EditTextAmountWatcher(etExchange))
        setSpinners()
        hideTouchOutsideManager.hideKeyboardByTouchOutsideEditText(scrollView, requireActivity())
    }
}